### ISSUE TITLE
Leaking file descriptors with UnixFd-containing messages

### ISSUE THREAD

**Issue by cmaves** [KillingSpark/rustbus#41]
Currently it is difficult to avoid leaking file descriptors when using a UnixFd enabled `RpcConn`. These unclosed file descriptors could prove problematic for long running processes Here are a few problem scenarios I have encountered.:
1) If a message is filtered out using `set_filter()`'s function, then any `RawFd`s in the filtered-out messages are leaked.
2) When processing an incoming method, if you receive an unexpected UnixFd and don't handle it by closing it you will leak the file descriptor. This possibility has to be handled for every message received and results in a lot of boilerplate code.
3) After sending a message that has UnixFds, you may wish to close them in your process. You use to be able to do this by iterating through the `raw_fds: Vec<RawFd>` but this is now private so this isn't possible anymore.

My proposed solution:
Change `MarshalledMessageBody.raw_fds` to `raw_fds: Vec<UnixFd>`. 
Change `UnixFd` to keep track of if it is owns the `RawFd` with `Cell`
If `UnixFd` owns `RawFd` then close it on drop.
```rust
struct UnixFdInner {
    fd: RawFd,
    owns: Cell<bool>
}
impl Drop for UnixFdInner {
   fn drop(&mut self) {
       if self.owns.get() {
           nix::unistd::close(self.fd).ok(); // silently ignore error if it fails to close
       }
  }
}
#[derive(Clone)]
pub struct UnixFd(Rc<UnixFdInner>);
impl UnixFd {
    pub fn new(fd: RawFd) -> Self {
        UnixFd(Rc::new(UnixFdInner {
            fd,
            owns: Cell::new(true)
        }))
    }
    /// Creates UnixFd that doesn't own RawFd, and thus doesn't 
    /// drop it when all are removed
    pub fn new_unowned(fd: RawFd) -> Self {
        UnixFd(Rc::new(UnixFdInner {
            fd,
            owns: Cell::new(false)
        }))
    }
}
impl AsRawFd for UnixFd {
    /// Retrieves the RawFd but doesn't take ownership
   fn as_raw_fd(&self) -> RawFd {
       self.0.fd
   }
}
impl IntoRawFd for UnixFd {
    /// Takes ownership away from UnixFd when getting RawFd.
    fn into_raw_fd(self) -> RawFd {
        self.0.owns.set(false);
        self.0.fd
    }
}
```
This `UnixFd` solves these problems by closing the descriptors if they are never accessed. The biggest downside to this is `UnixFd`, and `MarshalledMessageBody` (and by extension `MarshalledMessage`) would not longer be `Send` or `Sync`.

**Comment by KillingSpark** [comment]
This is true and very important, I did see some of this problem, but not the whole extent. I think marshal and unmarshal should use `dup()` to make a new fd. Then MarshalledMessage can `close()` all raw_fds in the drop impl, without 
1. leaking any fds (handled or not) in the MarshalledMessage
2. affecting any (un-)marshalled fields

I think this should solve all 3 problems you listed.

This would make the Rc<InnerUnixFd> though. If you use IntoRawFd you cannot use AsRawFd anyways, since it consumes `self`, and an "unowned" UnixFd would not exist in this case, because they are always owning, because they have a duped version.

This would mean having a new function like this:
```rust
impl UnixFd {
    pub fn new<T: AsRawFd>(fd: T) -> Result<Self, nix::Error> {
        Ok(UnixFd(nix::unistd::dup(fd.as_raw_fd())?))
    }
}
```

**Comment by cmaves** [comment]
`dup()` could work. It is a lot simpler which generally good. But I have concerns with it:
1) It could be a lot slower, because it introduces a syscall for every unmarshalled and marshalled `RawFd`; vs `Rc<UnixFdInner>` which consist of a memory allocation (on initial message reception only) and incrementing and decrementing a counter otherwise. Also a syscall introduce another unexpected point of failure when unmarshalling in the form of a syscall error.
There is also a possible additional `close()` that has to be done with the newly duplicated syscall.
2) It provides weaker guarantees against leaking file descriptors because unmarshalled doesn't necessarily mean accessed/handled. If somebody unmarshals into a `Param` and doesn't handle the possible UnixFd, or unmarshals into a collection such as a `HashMap<String, MarcroGeneratedVariantThatContainsUnixFd>` and doesn't check that an unexpected UnixFd isn't there. This is a corner case for sure that seems wouldn't normally be hit (without malicious intent), but Rust seems to be all about avoiding this kind of leak if possible.
3) Does a duped file descriptor count against the process file descriptor limit? If so, this may be undesirable. What if I were to send a DBus message with 1 file descriptor that is referenced 100,000 times within the message (I think DBus only limits the number of unique fds in the header, not how many times it is referenced?). Is this going to be duplicated with 100,000 calls to `dup()` and will this cause a OS limit to be reached? If these hypotheticals are true, then I feel like the are a potential security/DoS issue.

In order to avoid the second problem, I think `UnixFd` should handle cleaning up the file descriptor through a `Drop` impl, unless the user of the library *explicitly* takes ownership of the `RawFd` away from the `UnixFd`. This can be accomplished by giving each UnixFd a `dup`ed version or an `Rc`. If the `!Send` is of concern then you consider using `Arc` and `AtomicBool` instead. I'm not sure what kind of performance cost they ultimately have.

**Comment by cmaves** [comment]
So I did some testing and learned a few things:
1) Duped fds do count against the fd limit.
2) A combination of `dup()`/`close()` is about 100x more expensive than a `Rc::clone()`/`drop(Rc)` combo and 50x more expensive than an `Arc::clone()`/`drop(Arc)` combo in terms of time taken.

**Comment by KillingSpark** [comment]
Filedescriptor sending has always been a source of headache in terms of OS resource ddos attacks. Nothing prevents me from sending a message with the same FD a million times, which each would open a new FD in the receiving process. This is fine-ish since this can only happen if you have a malicious program running locally on your PC (and then you have a lot of other issues...). I get that this is a concern but I don't think that is really preventable.

As for the performance: I am not too worried with having FD-passing be a bit slower. This should not be any kind of high-throughput code path, I'd expect FDs to be passed only a few times between two processes. I'd much rather be able to rely on proven methods of having shared ownership of a resource pointed to by an FD, using the `dup()` call.

The weaker guarantees, you mention should be relatively easy to fix, by giving everything in the lib that contains an FD a drop impl that `close()`es it. This list is pretty small if I am not mistaken:
1. wire::marshal::traits::UnixFd
1. params::Base::UnixFd

So params::Base needs a Drop impl and wire::marshal::traits::UnixFd needs one too. 

This also needs to prevent them from closing FDs the user actually wants to use. 

1. They could have a Cell<Option<RawFd>> as the inner value. Then the user can take the FD out once, at which point they take responsibility of closing it properly, or the drop impl closes it if the cell holds a Some(fd) at drop time. 
1. They could require to be used with a `IntoRawFd` impl which would make sure that the inner FD is only taken out once.
1. They could have a 'get_duped()' function that `dup()`es the FD yet again but allows for accessing the FD multiple times

**Comment by cmaves** [comment]
If we go with `dup()` solution then I would go with use the 2nd option, where they are required to use the `IntoRawFd` to take ownership and have `param::Base::UnixFd` have contain a `UnixFd` instead. Then implement drop on `UnixFd` only. This is easier to keep track.

I still disagree with using `dup()`. A malicious or poorly written program on a single-user system is a lower risk, but for larger multi-user systems this threat is relevant and preventable. I also don't like adding a global side-effect to unmarshalling, to what is a currently self-contained operation. While `dup()` can provide multiple ownership, `Rc/Cell` avoids multiple ownership of the file descriptor altogether while being cheaper.

**Comment by KillingSpark** [comment]
I am convinced that using `dup` is the way to go for adding / getting fds from a MarshalledMessage. The message itself should control by it's own lifetime when the FDs it contains can be closed. Otherwise you might end up in weird states, where some of them have been closed by other means (like converting the fd to a std::fs::File and dropping that). It would (somewhat) bind the lifetime of the message to the lifetime of the contained FDs. Note that these could be closed and reused, while the message still thinks it holds the 'correct' FD.

This explicitly decouples the message from any other parts of the program, which seems very desirable. Then we can reason about only how UnixFd is handling the ownership without having to worry about how it could effect the message.

How UnixFd and Base::UnixFd behave, is more tricky. 
1. If you are using them to marshal stuff, you probably don't want them to take ownership and close the fd on drop. The FD is probably coming from another thing, that manages the lifetime of that FD. How this relationship is handled is probably not enforceable by rustbus.
1. If you are using them to unmarshal stuff, you definitely want them to take ownership at first. But you might want them to release ownership later without closing the FD at drop later. 

Your initial proposal might do the trick here, maybe without the AsRawFd, as that might lead people to write buggy code accidentally. Instead the `get_duped()` function could be provided, for cases where consuming the value is not an (easy) option.

I will sleep over this, and rethink this again tomorrow, but I am pretty sure `dup()ing` is the correct way to handle this. I see how this might be dangerous in terms of possible ddos attacks, but I also see how not doing it could result in weird, subtle bugs in applications using rustbus.

**Comment by cmaves** [comment]
I'll make a final case for against `dup()` and for `Rc` by raising one more issue and  leave it up to you.

File descriptors can be thought of as a pointer to one underlying kernel object that has interior mutability. Using `dup()` in this way is effectively a reference counter only in the kernel-space and not the user-space. You are doing the same thing just in a more expensive way. But it is worse because it gives the *illusion* of ownership (in the same way `Rc` does) without being able to tell the difference between two truly unique `RawFd`s or two `RawFd` that are actually the same kernel object. Using `dup()` opens the possibility of of creating two or more distinct objects like `std::fs::File` that refer to the same underlying object causing them to step all over each other and lead to some nasty bugs that violate Rust's rule against multiple mutable aliases. Now if you receive a message with multiple file descriptors you now have to make a syscall to validate that they aren't the same (I don't even know what syscall you would use to check if there the same.) 

You made a good observation that my proposed `Rc` solution could lead to a situation where you get a `UnixFd` that is already closed in the message an you wouldn't know this. In order to address this I would do change which is based on something you already suggested:
```rust
struct UnixFdInner {
    inner: Cell<Option<RawFd>>
}
impl Drop for UnixFdInner {
   fn drop(&mut self) {
       if let Some(fd) = self.inner.get() {
            nix::unistd::close(fd).ok();
       }
   }
}
#[derive(Clone)]
pub struct UnixFd(Rc<UnixFdInner>);
impl UnixFd {
    pub fn new(fd: RawFd) -> Self {
        UnixFd(Rc::new(UnixFdInner { inner: Cell::new(Some(fd)) }))
    }
    /// Gets a non-owning `RawFd`. If `None` is returned.
    /// then this UnixFd has already been taken by somebody else
    /// and is no longer valid.
    pub fn get_raw_fd(&self) -> Option<RawFd> {
        self.0.inner.get()
    }
    /// Gets a owning `RawFd` from the UnixFd. 
    /// Subsequent attempt to get the `RawFd` from
    /// other `UnixFd` referencing the same file descriptor will
    /// fail.
    pub fn take_raw_fd(self) -> Option<RawFd> {
        self.0.inner.take()
    }
}
impl PartialEq<UnixFd> for UnixFd {
    fn eq(&self, other: &UnixFd) -> bool {
        Rc::ptr_eq(&self.0, &other.0)
    }
}
```
Unfortunately using `UnixFd` in this way does mean that the unmarshalled message can be modified by using `get_raw_fd()/take_raw_fd()` which is kinda annoying but it informs programmer something is amiss with the received message, and is a necessary consequence because of the unsafe `RawFd` type. You **cannot** have multiple "owning" `RawFd`s that point to the same underlying object (regardless of whether it is a `dup()`ed `RawFd` or just a copied `RawFd`).  It asking for too much trouble.

**Comment by KillingSpark** [comment]
You make some very good points!
> File descriptors can be thought of as a pointer to one underlying kernel object that has interior mutability. Using dup() in this way is effectively a reference counter only in the kernel-space and not the user-space. You are doing the same thing just in a more expensive way. But it is worse because it gives the illusion of ownership (in the same way Rc does) without being able to tell the difference between two truly unique RawFds or two RawFd that are actually the same kernel object. Using dup() opens the possibility of of creating two or more distinct objects like std::fs::File that refer to the same underlying object causing them to step all over each other and lead to some nasty bugs that violate Rust's rule against multiple mutable aliases. Now if you receive a message with multiple file descriptors you now have to make a syscall to validate that they aren't the same (I don't even know what syscall you would use to check if they're the same.)

It seems like either way, there is no fool-proof solution to doing this. I thought `dup()ing` would cover more pitfalls then it creates but you're right, this is potentially way worse. 

What annoys me a little is that this code works the way it does:
```rust
let x = parser.get<UnixFd>().unwrap();
let y = parser.get<UnixFd>().unwrap();
y.get_fd();   // returns fd
x.take_fd();  // take ownership potentially somewhere far away in your codebase
y.get_fd();   // does not return the fd
```
But I guess that is inevitable (and likely not a problem, why would you do this anyways...). It just is a break in the otherwise clean separation between unmarshalled values.

I am starting to like the idea of the `UnixFd(InnerUnixFd)` more than my `dup()` version. This also very cleanly solves the issue that the `dup()` version had, for failed unmarshalling, where potentially a lot of `dup()` + `close()` would have been called. I think it is worth to spend a bit more time to think this through, but I guess you have me convinced.

**Comment by KillingSpark** [comment]
One last issue with the InnerUnixFd. This version always assumes ownership, which might be undesirable, if you are sending filedescriptors. This could be solved by having a separate type (basically the one existing in rustbus right now) that does not assume ownership of the fd, and that does indeed gets `dup()ed` (yeah I am sorry I am bringing this up as a solution yet again :D) when getting marshalled.

We could also leave the task of `dup()ing` the fd to the users, and notifiy them with a doc-comment of the consequences of using the UnixFd::new() for FDs that they want to keep around.

**Comment by KillingSpark** [comment]
I have started implementing this change in [this branch](https://github.com/KillingSpark/rustbus/tree/prevent_fd_leaking). The new UnixFd impl is in `rustbus/wire/unixfd.rs`. The params::UnixFd is not yet converted to using the UnixFd because it would need a Hash and Eq implementation, and I am not sure how to handle them.

**Comment by cmaves** [comment]
I dislike the `get_fd()` being different after `take_fd()` for the same reason as you pointed out. The reason it has to be this way is that `get_fd()` always needs to return a valid file descriptor. If a "owning" file descriptor is taken you cannot really know if a file descriptor is still valid or even worse, reused.

 I also dislike the `get_fd()` and `take_fd()` return the same thing. The differences is merely the semantics of the functions which is confusing. I am somewhat comforted in the fact standard library has this same issue with `AsRawFd` and `IntoRawFd`. 

Using `dup()` in some way is to allow creating new `UnixFd`s for marshalling, without giving up local ownership seems like a good idea.

**Comment by KillingSpark** [comment]
> Using dup() in some way is to allow creating new UnixFds for marshalling, without giving up local ownership seems like a good idea.

Glad to hear, then that is what I am gonna go with. It messes up the tests a bit, because now I have no real way of testing that the correct FD ends up in the message (there is a call in linux `kcmp` but it is not in rusts libc so I'll not bother for now...) it's fine-ish to just check that there is a new one, and the index got marshalled correctly.

I'll just need to figure out a sensible way to integrate the new UnixFd with params::Base::UnixFd, and then this should be pretty much done.

**Comment by KillingSpark** [comment]
Ok, I think this is pretty much ready to get merged into master. I'll have a final look over the changes I made soon-ish. If you have the time to have a look too that would be greatly appreciated.

**Comment by KillingSpark** [comment]
I merged this into master and published version 0.9.0 to crates.io

I think that should close this issue too. If you think there are still flaws, please feel free to reopen it :)

Thanks very much for contributing this fix for this very important problem!
